pkg/sql/alter_column_type_test.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql_test
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/tests"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+// TestInsertBeforeOldColumnIsDropped converts a column from INT to STRING
+// and then tries inserting 'hello' into the new column before the old column
+// has been dropped expecting a parse error on insert.
+func TestInsertBeforeOldColumnIsDropped(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	params, _ := tests.CreateTestServerParams()
+	params.Knobs = base.TestingKnobs{
+		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+			SkipComputedColumnSwapCleanup: true,
+		},
+	}
+
+	s, db, _ := serverutils.StartServer(t, params)
+	sqlDB := sqlutils.MakeSQLRunner(db)
+	ctx := context.Background()
+	defer s.Stopper().Stop(ctx)
+
+	expected := "pq: This table is still undergoing the ALTER COLUMN TYPE schema change, this insert is not supported until the schema change is finalized: could not parse \"hello\" as type int: strconv.ParseInt: parsing \"hello\": invalid syntax"
+
+	// This insert uses the insert fast path.
+	sqlDB.ExpectErr(t, expected, `
+SET enable_experimental_alter_column_type_general = true;
+CREATE TABLE test (x INT);
+ALTER TABLE test ALTER COLUMN x  TYPE STRING;
+INSERT INTO test VALUES ('hello');
+`)
+
+	expected = "pq: This table is still undergoing the ALTER COLUMN TYPE schema change, this insert may not be supported until the schema change is finalized: could not parse \"hello\" as type int: strconv.ParseInt: parsing \"hello\": invalid syntax"
```

Comment:
err is the same here right? you don't need to redefine it.

Reply:
The error is slightly different for the fast path insert vs non fast path insert. I put `may` be disabled due to ALTER COLUMN TYPE schema change for the non fast path insert because it could be an actual parse error - due to not actually being able to check the column.

Reply:
I see -- is it possible to just have this test search for the prefix of "this table is still undergoing ..."? I don't think there's a need for the whole string.